File: internal/services/bigquery/storage.go
package bigquery

import (
	"encoding/json"
	"errors"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

func (s *Server) datasetDir(projectID string, datasetID string) string {
	return filepath.Join(s.storageRoot(), "projects", projectID, "datasets", datasetID)
}

func (s *Server) datasetPath(projectID string, datasetID string) string {
	return filepath.Join(s.datasetDir(projectID, datasetID), "dataset.json")
}

func (s *Server) datasetIAMPolicyPath(projectID string, datasetID string) string {
	return filepath.Join(s.datasetDir(projectID, datasetID), "iam-policy.json")
}

func (s *Server) tableDir(projectID string, datasetID string, tableID string) string {
	return filepath.Join(s.datasetDir(projectID, datasetID), "tables", tableID)
}

func (s *Server) tablePath(projectID string, datasetID string, tableID string) string {
	return filepath.Join(s.tableDir(projectID, datasetID, tableID), "table.json")
}

func (s *Server) tableIAMPolicyPath(projectID string, datasetID string, tableID string) string {
	return filepath.Join(s.tableDir(projectID, datasetID, tableID), "iam-policy.json")
}

func (s *Server) routineDir(projectID string, datasetID string, routineID string) string {
	return filepath.Join(s.datasetDir(projectID, datasetID), "routines", routineID)
}

func (s *Server) routinePath(projectID string, datasetID string, routineID string) string {
	return filepath.Join(s.routineDir(projectID, datasetID, routineID), "routine.json")
}

func (s *Server) rowsPath(projectID string, datasetID string, tableID string) string {
	return filepath.Join(s.tableDir(projectID, datasetID, tableID), "rows", "streaming-buffer.jsonl")
}

func (s *Server) queryJobPath(projectID string, jobID string) string {
	return filepath.Join(s.storageRoot(), "projects", projectID, "jobs", jobID+".json")
}

func (s *Server) storageRoot() string {
	if strings.TrimSpace(s.config.StoragePath) == "" {
		return filepath.Join(".devcloud", "data", "bigquery")
	}
	return s.config.StoragePath
}

func (s *Server) defaultLocation() string {
	if strings.TrimSpace(s.config.Location) == "" {
		return "US"
	}
	return s.config.Location
}

func (s *Server) maxRequestBytes() int64 {
	if s.config.MaxRequestBytes <= 0 {
		return 10 * 1024 * 1024
	}
	return s.config.MaxRequestBytes
}

func (s *Server) maxRowsPerTable() int64 {
	if s.config.MaxRowsPerTable <= 0 {
		return 1000000
	}
	return s.config.MaxRowsPerTable
}

func (s *Server) maxResultRows() int {
	if s.config.MaxResultRows <= 0 {
		return 10000
	}
	return s.config.MaxResultRows
}

func (s *Server) effectiveUseLegacySQL(value *bool) bool {
	if value == nil {
		return s.config.DefaultLegacySQL
	}
	return *value
}

func (s *Server) readDataset(projectID string, datasetID string) (datasetResource, bool, error) {
	f, err := os.Open(s.datasetPath(projectID, datasetID))
	if errors.Is(err, os.ErrNotExist) {
		return datasetResource{}, false, nil
	}
	if err != nil {
		return datasetResource{}, false, err
	}
	defer f.Close()

	var dataset datasetResource
	if err := json.NewDecoder(f).Decode(&dataset); err != nil {
		return datasetResource{}, false, err
	}
	return dataset, true, nil
}

func (s *Server) readDatasets(projectID string) ([]datasetResource, error) {
	root := filepath.Join(s.storageRoot(), "projects", projectID, "datasets")
	entries, err := os.ReadDir(root)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	datasets := make([]datasetResource, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		dataset, found, err := s.readDataset(projectID, entry.Name())
		if err != nil {
			return nil, err
		}
		if found {
			datasets = append(datasets, dataset)
		}
	}
	sort.Slice(datasets, func(i, j int) bool {
		return datasets[i].DatasetReference.DatasetID < datasets[j].DatasetReference.DatasetID
	})
	return datasets, nil
}

func (s *Server) writeDataset(dataset datasetResource) error {
	path := s.datasetPath(dataset.DatasetReference.ProjectID, dataset.DatasetReference.DatasetID)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(f)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(dataset); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

func (s *Server) readIAMPolicy(path string) (iamPolicy, error) {
	f, err := os.Open(path)
	if errors.Is(err, os.ErrNotExist) {
		return defaultIAMPolicy(), nil
	}
	if err != nil {
		return iamPolicy{}, err
	}
	defer f.Close()

	var policy iamPolicy
	if err := json.NewDecoder(f).Decode(&policy); err != nil {
		return iamPolicy{}, err
	}
	return normalizeIAMPolicy(policy), nil
}

func (s *Server) writeIAMPolicy(path string, policy iamPolicy) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(f)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(policy); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

func (s *Server) readTable(projectID string, datasetID string, tableID string) (tableResource, bool, error) {
	f, err := os.Open(s.tablePath(projectID, datasetID, tableID))
	if errors.Is(err, os.ErrNotExist) {
		return tableResource{}, false, nil
	}
	if err != nil {
		return tableResource{}, false, err
	}
	defer f.Close()

	var table tableResource
	if err := json.NewDecoder(f).Decode(&table); err != nil {
		return tableResource{}, false, err
	}
	return table, true, nil
}

func (s *Server) readTables(projectID string, datasetID string) ([]tableResource, error) {
	root := filepath.Join(s.datasetDir(projectID, datasetID), "tables")
	entries, err := os.ReadDir(root)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	tables := make([]tableResource, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		table, found, err := s.readTable(projectID, datasetID, entry.Name())
		if err != nil {
			return nil, err
		}
		if found {
			tables = append(tables, table)
		}
	}
	sort.Slice(tables, func(i, j int) bool {
		return tables[i].TableReference.TableID < tables[j].TableReference.TableID
	})
	return tables, nil
}

func (s *Server) writeTable(table tableResource) error {
	path := s.tablePath(table.TableReference.ProjectID, table.TableReference.DatasetID, table.TableReference.TableID)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(f)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(table); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

func (s *Server) readRoutine(projectID string, datasetID string, routineID string) (routineResource, bool, error) {
	f, err := os.Open(s.routinePath(projectID, datasetID, routineID))
	if errors.Is(err, os.ErrNotExist) {
		return routineResource{}, false, nil
	}
	if err != nil {
		return routineResource{}, false, err
	}
	defer f.Close()

	var routine routineResource
	if err := json.NewDecoder(f).Decode(&routine); err != nil {
		return routineResource{}, false, err
	}
	return routine, true, nil
}

func (s *Server) readRoutines(projectID string, datasetID string) ([]routineResource, error) {
	root := filepath.Join(s.datasetDir(projectID, datasetID), "routines")
	entries, err := os.ReadDir(root)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	routines := make([]routineResource, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		routine, found, err := s.readRoutine(projectID, datasetID, entry.Name())
		if err != nil {
			return nil, err
		}
		if found {
			routines = append(routines, routine)
		}
	}
	sort.Slice(routines, func(i, j int) bool {
		return routines[i].RoutineReference.RoutineID < routines[j].RoutineReference.RoutineID
	})
	return routines, nil
}

func (s *Server) writeRoutine(routine routineResource) error {
	path := s.routinePath(routine.RoutineReference.ProjectID, routine.RoutineReference.DatasetID, routine.RoutineReference.RoutineID)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(f)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(routine); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

func (s *Server) appendRows(projectID string, datasetID string, tableID string, rows []storedRow) error {
	path := s.rowsPath(projectID, datasetID, tableID)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(f)
	for _, row := range rows {
		if err := encoder.Encode(row); err != nil {
			f.Close()
			return err
		}
	}
	return f.Close()
}

func (s *Server) readRows(projectID string, datasetID string, tableID string) ([]storedRow, error) {
	f, err := os.Open(s.rowsPath(projectID, datasetID, tableID))
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	defer f.Close()

	decoder := json.NewDecoder(f)
	var rows []storedRow
	for {
		var row storedRow
		if err := decoder.Decode(&row); errors.Is(err, io.EOF) {
			break
		} else if err != nil {
			return nil, err
		}
		rows = append(rows, row)
	}
	return rows, nil
}

func (s *Server) writeQueryJob(projectID string, jobID string, job queryJobRecord) error {
	path := s.queryJobPath(projectID, jobID)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	tmp := path + ".tmp"
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	encoder := json.NewEncoder(f)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(job); err != nil {
		f.Close()
		os.Remove(tmp)
		return err
	}
	if err := f.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	return os.Rename(tmp, path)
}

func (s *Server) readQueryJob(projectID string, jobID string) (queryJobRecord, bool, error) {
	f, err := os.Open(s.queryJobPath(projectID, jobID))
	if errors.Is(err, os.ErrNotExist) {
		return queryJobRecord{}, false, nil
	}
	if err != nil {
		return queryJobRecord{}, false, err
	}
	defer f.Close()

	var job queryJobRecord
	if err := json.NewDecoder(f).Decode(&job); err != nil {
		return queryJobRecord{}, false, err
	}
	return job, true, nil
}

func (s *Server) readQueryJobs(projectID string) ([]JobSnapshot, error) {
	records, err := s.readQueryJobRecords(projectID)
	if err != nil {
		return nil, err
	}
	jobs := make([]JobSnapshot, 0, len(records))
	for _, job := range records {
		jobs = append(jobs, jobSnapshotFromRecord(job))
	}
	return jobs, nil
}

func (s *Server) readQueryJobRecords(projectID string) ([]queryJobRecord, error) {
	root := filepath.Join(s.storageRoot(), "projects", projectID, "jobs")
	entries, err := os.ReadDir(root)
	if errors.Is(err, os.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	jobs := make([]queryJobRecord, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".json") {
			continue
		}
		jobID := strings.TrimSuffix(entry.Name(), ".json")
		job, found, err := s.readQueryJob(projectID, jobID)
		if err != nil {
			return nil, err
		}
		if found {
			jobs = append(jobs, job)
		}
	}
	sort.Slice(jobs, func(i, j int) bool {
		return jobs[i].Job.JobReference.JobID < jobs[j].Job.JobReference.JobID
	})
	return jobs, nil
}

func jobSnapshotFromRecord(job queryJobRecord) JobSnapshot {
	return JobSnapshot{
		ProjectID: job.Job.JobReference.ProjectID,
		JobID:     job.Job.JobReference.JobID,
		Location:  job.Job.JobReference.Location,
		State:     job.Job.Status.State,
		Job:       job.Job,
	}
}

func rawMapForSnapshot(values map[string]json.RawMessage) map[string]any {
	result := make(map[string]any, len(values))
	for key, raw := range values {
		var value any
		decoder := json.NewDecoder(strings.NewReader(string(raw)))
		decoder.UseNumber()
		if err := decoder.Decode(&value); err != nil {
			result[key] = string(raw)
			continue
		}
		result[key] = value
	}
	return result
}

func (s *Server) refreshTableRowStats(table tableResource) error {
	rows, err := s.readRows(table.TableReference.ProjectID, table.TableReference.DatasetID, table.TableReference.TableID)
	if err != nil {
		return err
	}
	var bytes int
	for _, row := range rows {
		data, err := json.Marshal(row.JSON)
		if err != nil {
			return err
		}
		bytes += len(data)
	}
	now := time.Now().UTC()
	table.NumRows = strconv.Itoa(len(rows))
	table.NumBytes = strconv.Itoa(bytes)
	table.ETag = datasetETag(now)
	table.LastModifiedTime = unixMillisString(now)
	return s.writeTable(table)
}

File: services/bigquery/src/storage.rs
//! On-disk persistence — port of `internal/services/bigquery/storage.rs`.
//!
//! Layout under the storage root (default `.devcloud/data/bigquery`), byte
//! compatible with the legacy service so state survives switching engines:
//!
//! ```text
//! projects/<project>/datasets/<dataset>/dataset.json          (indent 2 + \n)
//! projects/<project>/datasets/<dataset>/iam-policy.json       (indent 2 + \n)
//! projects/<project>/datasets/<dataset>/tables/<t>/table.json (indent 2 + \n)
//! projects/<project>/datasets/<dataset>/tables/<t>/iam-policy.json
//! projects/<project>/datasets/<dataset>/tables/<t>/rows/streaming-buffer.jsonl
//! projects/<project>/datasets/<dataset>/routines/<r>/routine.json
//! projects/<project>/jobs/<job>.json                          (indent 2 + \n)
//! ```
//!
//! Resource files are written atomically (`<path>.tmp` + rename), exactly like
//! the legacy implementation; the streaming buffer is append-only JSONL with one
//! compact `json.Encoder`-style row per line.

use std::collections::BTreeMap;
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

use crate::model::{
    DatasetResource, IamPolicy, QueryJobRecord, RawJson, RoutineResource, StoredRow, TableResource,
};
use crate::responses::{
    dataset_etag, default_iam_policy, normalize_iam_policy, unix_millis_string,
};
use crate::server::{now_unix_nanos, Server};
use crate::wire_json;

/// Persistence failure: any I/O or JSON error maps to the legacy handlers' 500
/// `backendError`.
#[derive(Debug)]
pub enum StorageError {
    Io(std::io::Error),
    Json(serde_json::Error),
}

impl From<std::io::Error> for StorageError {
    fn from(err: std::io::Error) -> Self {
        StorageError::Io(err)
    }
}

impl From<serde_json::Error> for StorageError {
    fn from(err: serde_json::Error) -> Self {
        StorageError::Json(err)
    }
}

impl std::fmt::Display for StorageError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StorageError::Io(err) => write!(f, "{err}"),
            StorageError::Json(err) => write!(f, "{err}"),
        }
    }
}

pub type StorageResult<T> = Result<T, StorageError>;

impl Server {
    // --- paths (legacy storage.rs path helpers) --------------------------------

    pub fn storage_root(&self) -> PathBuf {
        if self.config.storage_path.trim().is_empty() {
            Path::new(".devcloud").join("data").join("bigquery")
        } else {
            PathBuf::from(&self.config.storage_path)
        }
    }

    pub fn dataset_dir(&self, project_id: &str, dataset_id: &str) -> PathBuf {
        self.storage_root()
            .join("projects")
            .join(project_id)
            .join("datasets")
            .join(dataset_id)
    }

    pub fn dataset_path(&self, project_id: &str, dataset_id: &str) -> PathBuf {
        self.dataset_dir(project_id, dataset_id)
            .join("dataset.json")
    }

    pub fn dataset_iam_policy_path(&self, project_id: &str, dataset_id: &str) -> PathBuf {
        self.dataset_dir(project_id, dataset_id)
            .join("iam-policy.json")
    }

    pub fn table_dir(&self, project_id: &str, dataset_id: &str, table_id: &str) -> PathBuf {
        self.dataset_dir(project_id, dataset_id)
            .join("tables")
            .join(table_id)
    }

    pub fn table_path(&self, project_id: &str, dataset_id: &str, table_id: &str) -> PathBuf {
        self.table_dir(project_id, dataset_id, table_id)
            .join("table.json")
    }

    pub fn table_iam_policy_path(
        &self,
        project_id: &str,
        dataset_id: &str,
        table_id: &str,
    ) -> PathBuf {
        self.table_dir(project_id, dataset_id, table_id)
            .join("iam-policy.json")
    }

    pub fn routine_dir(&self, project_id: &str, dataset_id: &str, routine_id: &str) -> PathBuf {
        self.dataset_dir(project_id, dataset_id)
            .join("routines")
            .join(routine_id)
    }

    pub fn routine_path(&self, project_id: &str, dataset_id: &str, routine_id: &str) -> PathBuf {
        self.routine_dir(project_id, dataset_id, routine_id)
            .join("routine.json")
    }

    pub fn rows_path(&self, project_id: &str, dataset_id: &str, table_id: &str) -> PathBuf {
        self.table_dir(project_id, dataset_id, table_id)
            .join("rows")
            .join("streaming-buffer.jsonl")
    }

    pub fn query_job_path(&self, project_id: &str, job_id: &str) -> PathBuf {
        self.storage_root()
            .join("projects")
            .join(project_id)
            .join("jobs")
            .join(format!("{job_id}.json"))
    }

    // --- datasets -----------------------------------------------------------

    pub fn read_dataset(
        &self,
        project_id: &str,
        dataset_id: &str,
    ) -> StorageResult<Option<DatasetResource>> {
        read_json_resource(&self.dataset_path(project_id, dataset_id))
    }

    pub fn read_datasets(&self, project_id: &str) -> StorageResult<Vec<DatasetResource>> {
        let root = self
            .storage_root()
            .join("projects")
            .join(project_id)
            .join("datasets");
        let mut datasets: Vec<DatasetResource> = Vec::new();
        for name in read_subdir_names(&root)? {
            if let Some(dataset) = self.read_dataset(project_id, &name)? {
                datasets.push(dataset);
            }
        }
        datasets.sort_by(|a, b| {
            a.dataset_reference
                .dataset_id
                .cmp(&b.dataset_reference.dataset_id)
        });
        Ok(datasets)
    }

    pub fn write_dataset(&self, dataset: &DatasetResource) -> StorageResult<()> {
        let path = self.dataset_path(
            &dataset.dataset_reference.project_id,
            &dataset.dataset_reference.dataset_id,
        );
        write_atomic(&path, &wire_json::to_vec_indent(dataset))
    }

    // --- IAM policies -------------------------------------------------------

    pub fn read_iam_policy(&self, path: &Path) -> StorageResult<IamPolicy> {
        let data = match fs::read(path) {
            Ok(data) => data,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
                return Ok(default_iam_policy())
            }
            Err(err) => return Err(err.into()),
        };
        let policy: IamPolicy = serde_json::from_slice(&data)?;
        Ok(normalize_iam_policy(policy, now_unix_nanos()))
    }

    pub fn write_iam_policy(&self, path: &Path, policy: &IamPolicy) -> StorageResult<()> {
        write_atomic(path, &wire_json::to_vec_indent(policy))
    }

    // --- tables -------------------------------------------------------------

    pub fn read_table(
        &self,
        project_id: &str,
        dataset_id: &str,
        table_id: &str,
    ) -> StorageResult<Option<TableResource>> {
        read_json_resource(&self.table_path(project_id, dataset_id, table_id))
    }

    pub fn read_tables(
        &self,
        project_id: &str,
        dataset_id: &str,
    ) -> StorageResult<Vec<TableResource>> {
        let root = self.dataset_dir(project_id, dataset_id).join("tables");
        let mut tables: Vec<TableResource> = Vec::new();
        for name in read_subdir_names(&root)? {
            if let Some(table) = self.read_table(project_id, dataset_id, &name)? {
                tables.push(table);
            }
        }
        tables.sort_by(|a, b| a.table_reference.table_id.cmp(&b.table_reference.table_id));
        Ok(tables)
    }

    pub fn write_table(&self, table: &TableResource) -> StorageResult<()> {
        let path = self.table_path(
            &table.table_reference.project_id,
            &table.table_reference.dataset_id,
            &table.table_reference.table_id,
        );
        write_atomic(&path, &wire_json::to_vec_indent(table))
    }

    // --- routines -----------------------------------------------------------

    pub fn read_routine(
        &self,
        project_id: &str,
        dataset_id: &str,
        routine_id: &str,
    ) -> StorageResult<Option<RoutineResource>> {
        read_json_resource(&self.routine_path(project_id, dataset_id, routine_id))
    }

    pub fn read_routines(
        &self,
        project_id: &str,
        dataset_id: &str,
    ) -> StorageResult<Vec<RoutineResource>> {
        let root = self.dataset_dir(project_id, dataset_id).join("routines");
        let mut routines: Vec<RoutineResource> = Vec::new();
        for name in read_subdir_names(&root)? {
            if let Some(routine) = self.read_routine(project_id, dataset_id, &name)? {
                routines.push(routine);
            }
        }
        routines.sort_by(|a, b| {
            a.routine_reference
                .routine_id
                .cmp(&b.routine_reference.routine_id)
        });
        Ok(routines)
    }

    pub fn write_routine(&self, routine: &RoutineResource) -> StorageResult<()> {
        let path = self.routine_path(
            &routine.routine_reference.project_id,
            &routine.routine_reference.dataset_id,
            &routine.routine_reference.routine_id,
        );
        write_atomic(&path, &wire_json::to_vec_indent(routine))
    }

    // --- streaming buffer ---------------------------------------------------

    /// Appends rows to the streaming buffer, one compact `json.Encoder`-style
    /// line per row. Raw values are compacted on write, exactly as legacy
    /// encoder does for `json.RawMessage`.
    pub fn append_rows(
        &self,
        project_id: &str,
        dataset_id: &str,
        table_id: &str,
        rows: &[StoredRow],
    ) -> StorageResult<()> {
        let path = self.rows_path(project_id, dataset_id, table_id);
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }
        let mut file = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&path)?;
        for row in rows {
            file.write_all(&wire_json::to_vec(&compact_row(row)?))?;
        }
        Ok(())
    }

    pub fn read_rows(
        &self,
        project_id: &str,
        dataset_id: &str,
        table_id: &str,
    ) -> StorageResult<Vec<StoredRow>> {
        let data = match fs::read(self.rows_path(project_id, dataset_id, table_id)) {
            Ok(data) => data,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
            Err(err) => return Err(err.into()),
        };
        let mut rows = Vec::new();
        for row in serde_json::Deserializer::from_slice(&data).into_iter::<StoredRow>() {
            rows.push(row?);
        }
        Ok(rows)
    }

    // --- query jobs ---------------------------------------------------------

    pub fn write_query_job(
        &self,
        project_id: &str,
        job_id: &str,
        job: &QueryJobRecord,
    ) -> StorageResult<()> {
        write_atomic(
            &self.query_job_path(project_id, job_id),
            &wire_json::to_vec_indent(job),
        )
    }

    pub fn read_query_job(
        &self,
        project_id: &str,
        job_id: &str,
    ) -> StorageResult<Option<QueryJobRecord>> {
        read_json_resource(&self.query_job_path(project_id, job_id))
    }

    /// legacy `readQueryJobRecords`: every `<job>.json` under the project's jobs
    /// dir, sorted by job id.
    pub fn read_query_job_records(&self, project_id: &str) -> StorageResult<Vec<QueryJobRecord>> {
        let root = self
            .storage_root()
            .join("projects")
            .join(project_id)
            .join("jobs");
        let entries = match fs::read_dir(&root) {
            Ok(entries) => entries,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
            Err(err) => return Err(err.into()),
        };
        let mut jobs: Vec<QueryJobRecord> = Vec::new();
        for entry in entries {
            let entry = entry?;
            let name = entry.file_name().to_string_lossy().into_owned();
            if entry.file_type()?.is_dir() || !name.ends_with(".json") {
                continue;
            }
            let job_id = name.trim_end_matches(".json");
            if let Some(job) = self.read_query_job(project_id, job_id)? {
                jobs.push(job);
            }
        }
        jobs.sort_by(|a, b| a.job.job_reference.job_id.cmp(&b.job.job_reference.job_id));
        Ok(jobs)
    }

    // --- row stats ----------------------------------------------------------

    /// legacy `refreshTableRowStats`: recomputes `numRows`/`numBytes` (bytes =
    /// `len(json.Marshal(row.JSON))` summed) and bumps etag/lastModifiedTime.
    pub fn refresh_table_row_stats(&self, table: &TableResource) -> StorageResult<()> {
        let rows = self.read_rows(
            &table.table_reference.project_id,
            &table.table_reference.dataset_id,
            &table.table_reference.table_id,
        )?;
        let mut bytes = 0usize;
        for row in &rows {
            bytes += wire_json::marshal(&row.json).len();
        }
        let now = now_unix_nanos();
        let mut table = table.clone();
        table.num_rows = rows.len().to_string();
        table.num_bytes = bytes.to_string();
        table.etag = dataset_etag(now);
        table.last_modified_time = unix_millis_string(now);
        self.write_table(&table)
    }
}

/// Reads one JSON resource file; missing file → `None` (legacy
/// `(resource, false, nil)`).
fn read_json_resource<T: serde::de::DeserializeOwned>(path: &Path) -> StorageResult<Option<T>> {
    let data = match fs::read(path) {
        Ok(data) => data,
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(None),
        Err(err) => return Err(err.into()),
    };
    Ok(Some(serde_json::from_slice(&data)?))
}

/// Lists immediate subdirectory names. The legacy code iterates `os.ReadDir`
/// output (sorted by filename) and skips non-directories; the final ordering
/// comes from the explicit resource-id sort in each caller.
fn read_subdir_names(root: &Path) -> StorageResult<Vec<String>> {
    let entries = match fs::read_dir(root) {
        Ok(entries) => entries,
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(Vec::new()),
        Err(err) => return Err(err.into()),
    };
    let mut names = Vec::new();
    for entry in entries {
        let entry = entry?;
        if entry.file_type()?.is_dir() {
            names.push(entry.file_name().to_string_lossy().into_owned());
        }
    }
    names.sort();
    Ok(names)
}

/// legacy tmp-file + rename atomic write (`MkdirAll` + `O_CREATE|O_TRUNC` +
/// `Rename`).
fn write_atomic(path: &Path, data: &[u8]) -> StorageResult<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let tmp = path.with_extension(match path.extension() {
        Some(ext) => format!("{}.tmp", ext.to_string_lossy()),
        None => "tmp".to_string(),
    });
    if let Err(err) = fs::write(&tmp, data) {
        let _ = fs::remove_file(&tmp);
        return Err(err.into());
    }
    if let Err(err) = fs::rename(&tmp, path) {
        let _ = fs::remove_file(&tmp);
        return Err(err.into());
    }
    Ok(())
}

/// Re-creates a row with every raw JSON value compacted, matching what legacy
/// `json.Encoder` does to `json.RawMessage` on write.
fn compact_row(row: &StoredRow) -> StorageResult<StoredRow> {
    let mut json: BTreeMap<String, RawJson> = BTreeMap::new();
    for (key, value) in &row.json {
        json.insert(
            key.clone(),
            serde_json::value::RawValue::from_string(wire_json::compact(value.get()))?,
        );
    }
    Ok(StoredRow {
        insert_id: row.insert_id.clone(),
        json,
        inserted_at: row.inserted_at.clone(),
    })
}
